app/src/contexts/observationsContext.tsx

```diff
@@ -0,0 +1,189 @@
+  samplingSite: string | undefined;
+  samplingMethod: string | undefined;
+  samplingPeriod: string | undefined;
+  count: number | undefined;
+  observation_date: Date | undefined;
+  observation_time: string | undefined;
+  latitude: number | undefined;
+  longitude: number | undefined;
+}
+
+export interface IObservationTableRow extends Partial<IObservationRecord> {
+  id: string;
+}
+
+/**
+ * Context object that stores information about survey observations
+ *
+ * @export
+ * @interface IObservationsContext
+ */
+export type IObservationsContext = {
+  /**
+   * Appends a new blank record to the observation rows
+   */
+  createNewRecord: () => void;
+  /**
+   * Commits all observation rows to the database, including those that are currently being
+   * edited in the Observation Table
+   */
+  saveRecords: () => Promise<void>;
+  /**
+   * Reverts all changes made to observation records within the Observation Table
+   */
+  revertRecords: () => Promise<void>;
+  /**
+   * Refreshes the Observation Table with already existing records
+   */
+  refreshRecords: () => Promise<void>;
+  /**
+   * Marks the given record as unsaved
+   */
+  markRecordWithUnsavedChanges: (id: string) => void;
+  /**
+   * Indicates all observation table rows that have unsaved changes, include IDs of rows
+   * that have been deleted.
+   */
+  unsavedRecordIds: string[];
+  /**
+   * Inidicates whether the observation table has any unsaved changes
+   */
+  hasUnsavedChanges: () => boolean;
+  /**
+   * Data Loader used for retrieving existing records
+   */
+  observationsDataLoader: DataLoader<[], IGetSurveyObservationsResponse, unknown>;
+  /**
+   * API ref used to interface with an MUI DataGrid representing the observation records
+   */
+  _muiDataGridApiRef: React.MutableRefObject<GridApiCommunity>;
+};
+
+export const ObservationsContext = createContext<IObservationsContext>({
+  _muiDataGridApiRef: { current: null as unknown as GridApiCommunity },
+  observationsDataLoader: {} as DataLoader<never, IGetSurveyObservationsResponse, unknown>,
+  unsavedRecordIds: [],
+  markRecordWithUnsavedChanges: () => {},
+  hasUnsavedChanges: () => false,
+  createNewRecord: () => {},
+  revertRecords: () => Promise.resolve(),
+  saveRecords: () => Promise.resolve(),
+  refreshRecords: () => Promise.resolve(),
+});
+
+export const ObservationsContextProvider = (props: PropsWithChildren<Record<never, any>>) => {
+  const _muiDataGridApiRef = useGridApiRef();
+  const biohubApi = useBiohubApi();
+  const surveyContext = useContext(SurveyContext);
+  const { projectId, surveyId } = useContext(SurveyContext);
+  const observationsDataLoader = useDataLoader(() => biohubApi.observation.getObservationRecords(projectId, surveyId));
+  const [unsavedRecordIds, _setUnsavedRecordIds] = useState<string[]>([]);
+
+  observationsDataLoader.load();
+
+  const markRecordWithUnsavedChanges = (id: string) => {
+    const unsavedRecordSet = new Set<string>([...unsavedRecordIds]);
```

Comment:
Don't think this spread is needed

Reply:
I usually clone arrays when passing into a consturctor, but I agree; it's not needed.